File: LinkedList.java
public class LinkedList {
    private static class Node {
        int data;
        Node next;
        Node(int d) { data = d; }
    }

    private Node head;

    public LinkedList() { head = null; }

    public void create(int[] vals) {
        head = null;
        for (int v : vals) insert(v);
    }

    public void insert(int value) {
        Node node = new Node(value);
        if (head == null) { head = node; return; }
        Node cur = head;
        while (cur.next != null) cur = cur.next;
        cur.next = node;
    }

    public boolean delete(int value) {
        Node cur = head, prev = null;
        while (cur != null) {
            if (cur.data == value) {
                if (prev == null) head = cur.next;
                else prev.next = cur.next;
                return true;
            }
            prev = cur;
            cur = cur.next;
        }
        return false;
    }

    public void traverse() {
        System.out.print("[");
        Node cur = head;
        boolean first = true;
        while (cur != null) {
            if (!first) System.out.print(", ");
            System.out.print(cur.data);
            first = false;
            cur = cur.next;
        }
        System.out.println("]");
    }

    public static void main(String[] args) {
        LinkedList ll = new LinkedList();
        ll.create(new int[]{1,2,3});
        System.out.print("After create: "); ll.traverse();
        ll.insert(4);
        System.out.print("After insert 4: "); ll.traverse();
        boolean removed = ll.delete(2);
        System.out.println("Deleted 2? " + (removed ? "yes" : "no"));
        System.out.print("Final list: "); ll.traverse();
    }
}

File: linked_list.pas
program LinkedListDemo;

type
  PNode = ^TNode;
  TNode = record
    title: string;
    author: string;
    rating: Real;
    next: PNode;
  end;

procedure CreateList(var head: PNode);
begin
  head := nil;
end;

procedure InsertBook(var head: PNode; t, a: string; r: Real);
var
  node, cur: PNode;
begin
  New(node);
  node^.title := t;
  node^.author := a;
  node^.rating := r;
  node^.next := nil;
  if head = nil then
    head := node
  else
  begin
    cur := head;
    while cur^.next <> nil do
      cur := cur^.next;
    cur^.next := node;
  end;
end;

function DeleteBook(var head: PNode; t: string): Boolean;
var
  cur, prev: PNode;
begin
  cur := head;
  prev := nil;
  while cur <> nil do
  begin
    if cur^.title = t then
    begin
      if prev = nil then
        head := cur^.next
      else
        prev^.next := cur^.next;
      Dispose(cur);
      Exit(True);
    end;
    prev := cur;
    cur := cur^.next;
  end;
  DeleteBook := False;
end;

procedure TraverseBooks(head: PNode);
var
  cur: PNode;
  i: Integer;
begin
  cur := head;
  i := 1;
  if cur = nil then
  begin
    Writeln('(no books in the list)');
    Exit;
  end;
  while cur <> nil do
  begin
    Write(i); Write('. '); Write(cur^.title); Write(' by '); Write(cur^.author); Write(' - Rating: ');
    Write(cur^.rating:0:1);
    Writeln;
    cur := cur^.next;
    i := i + 1;
  end;
end;

var
  head: PNode;

begin
  CreateList(head);

  { sample books }
  InsertBook(head, 'The Hobbit', 'J.R.R. Tolkien', 4.8);
  InsertBook(head, '1984', 'George Orwell', 4.6);
  InsertBook(head, 'To Kill a Mockingbird', 'Harper Lee', 4.7);

  Writeln('Books after create:');
  TraverseBooks(head);

  InsertBook(head, 'Dune', 'Frank Herbert', 4.5);
  Writeln;
  Writeln('After inserting ''Dune'':');
  TraverseBooks(head);

  if DeleteBook(head, '1984') then
    Writeln;
    Writeln('Deleted ''1984'': yes')
  else
    Writeln('Deleted ''1984'': no');

  Writeln;
  Writeln('Final list:');
  TraverseBooks(head);
end.
